src/herder/test/UpgradesTests.cpp

```diff
@@ -1029,6 +1029,301 @@ TEST_CASE("SCP timing config affects consensus behavior", "[upgrades][herder]")
+    };
+
+    auto expectSingleValueStateSizeWindow =
+        [&](uint64_t value,
+            std::optional<uint32_t> expectedWindowSize = std::nullopt) {
+            if (!expectedWindowSize)
+            {
+                expectedWindowSize = windowSize;
+            }
+            std::vector<uint64_t> expectedWindow(*expectedWindowSize);
+            expectedWindow.assign(*expectedWindowSize, value);
+            REQUIRE(getStateSizeWindow() == expectedWindow);
+        };
+
+    auto const initBlSize =
+        test.getApp().getBucketManager().getLiveBucketList().getSize();
+
+    INFO("snapshot BL size in p22");
+    // Generate txs to fill up the state size window.
+    generateTxs(windowSize * samplePeriod * 2);
+
+    // We're still in p22, so the last snapshot must still be BL size.
+    auto const blSize =
+        test.getApp().getBucketManager().getLiveBucketList().getSize();
+    auto const p22StateSizeWindow = getStateSizeWindow();
+    verifyAverageStateSize(initBlSize, blSize);
+
+    // The BL grows by the updated config entry after we create the
+    // snapshot. That's why the actual BL size is a bit smaller than the
+    // snapshotted value.
+    int64_t blSizeDiff =
+        std::abs(static_cast<int64_t>(blSize) -
+                 static_cast<int64_t>(p22StateSizeWindow.back()));
+    REQUIRE(blSizeDiff <= 200);
+
+    {
+        INFO("track in-memory size in p22");
+        verifyExpectedInMemorySize(0);
+    }
+
+    {
+        INFO("perform settings upgrade in p22");
+        modifySorobanNetworkConfig(
+            test.getApp(), [](SorobanNetworkConfig& cfg) {
+                cfg.mMemCostParams[ContractCostType::ParseWasmInstructions]
+                    .constTerm *= 2;
+                cfg.mMemCostParams[ContractCostType::ParseWasmInstructions]
+                    .linearTerm *= 2;
+            });
+        ensureInMemorySizeIncreased();
+        // There is a small expected state size diff due to the settings upgrade
+        // contract.
+        verifyExpectedInMemorySize(100'000);
+
+        // The state size window must be unchanged.
+        REQUIRE(getStateSizeWindow() == p22StateSizeWindow);
+    }
+
+    INFO("upgrade to p23");
+    executeUpgrade(test.getApp(), makeProtocolVersionUpgrade(23));
+    // In-memory size shouldn't have changed as it has been computed with p23
+    // logic.
+    REQUIRE(test.getApp().getLedgerManager().getSorobanInMemoryStateSize() ==
+            lastInMemorySize);
+    auto const p23MemorySize = lastInMemorySize;
+    // State size window now contains only the current in-memory size.
+    expectSingleValueStateSizeWindow(p23MemorySize);
+    verifyAverageStateSize(p23MemorySize, p23MemorySize);
+
+    {
+        INFO("fill window with in-memory size in p23");
+        closeLedger(test.getApp());
+
+        // Now generate more txs to fill up the window with in-memory sizes.
+        generateTxs(windowSize * samplePeriod * 4);
+        verifyExpectedInMemorySize(0);
```

Comment:
Is maxDiff 0 here because the original upgrade has expired and been evicted?

Reply:
No, it's because we can estimate the expected size correctly when we just generate the transactions. The discrepancies only appear on settings upgrade, because it writes a bunch of data that we're not accounting for.